aea/protocols/dialogue/base.py

```diff
@@ -857,66 +851,119 @@ def _basic_validation_non_initial_message(
+            return "Invalid target {}. target_message can not be found.".format(target)
+
         target_performative = target_message.performative
         if performative not in self.rules.get_valid_replies(target_performative):
-            return (
-                False,
-                "Invalid performative. Expected one of {}. Found {}.".format(
-                    self.rules.get_valid_replies(target_performative), performative
-                ),
+            return "Invalid performative. Expected one of {}. Found {}.".format(
+                self.rules.get_valid_replies(target_performative), performative
             )
 
-        return True, "The non-initial message passes basic validation."
+        return None
 
-    def _additional_validation(self, message: Message) -> Tuple[bool, str]:
-        """
-        Check whether 'message' is a valid next message in the dialogue, according to additional rules.
+    def _validate_message_id(self, message: Message) -> Optional[str]:
+        """Check message id corresponds to message id sequences, if not return error string."""
+        is_outgoing = message.to != self.self_address
 
-        These rules are designed to be less fundamental than basic rules and subject to change.
-        Currently the following is enforced:
+        # This assumes that messages sent by the opponent are sent in the right order.
+        if is_outgoing:
+            next_message_id = self.get_outgoing_next_message_id()
+        else:
+            next_message_id = self.get_incoming_next_message_id()
 
-         - A message targets the message strictly before it in the dialogue
+        # we know what is the next message id for incoming and outgoing!
+        if message.message_id != next_message_id:
+            return "Invalid message_id. Expected {}. Found {}.".format(
+                next_message_id, message.message_id
+            )
 
-        :param message: the message to be validated
-        :return: Boolean result, and associated message.
-        """
-        if self.is_empty:
-            return True, "The message passes additional validation."
+        return None
 
-        last_target = self.last_message.target  # type: ignore
-        if message.target == last_target + 1:
-            return True, "The message passes additional validation."
+    def iter_all_messages(self) -> Iterable:
+        """Iterate over all messages in the dialogue."""
+        return chain(self._incoming_messages, self._outgoing_messages)
 
-        return (
-            False,
-            "Invalid target. Expected {}. Found {}.".format(
-                last_target + 1, message.target
-            ),
-        )
+    def get_message_by_id(self, message_id: int) -> Optional[Message]:
+        """Get message by id, if not presents return None."""
+        for message in self.iter_all_messages():
+            if message.message_id == message_id:
+                return message
+        return None
```

Comment:
This isn't efficient. No need to loop through all messages. Just access the `message_id-1`th message of incoming/outgoing_messages respecting the positive or negative message_id. 

We know messages in say outgoing (or incoming_messages) list preserves their order with respect to other outgoing (incoming) messages.

Reply:
updated